geo/src/algorithm/relate/geomgraph/intersection_matrix.rs

```diff
@@ -268,6 +278,212 @@ impl IntersectionMatrix {
+        // [F**T*****]
+        self.0[CoordPos::Inside][CoordPos::Inside] == Dimensions::Empty
+        && self.0[CoordPos::OnBoundary][CoordPos::Inside] != Dimensions::Empty ||
+        // [F***T****]
+        self.0[CoordPos::Inside][CoordPos::Inside] == Dimensions::Empty
+        && self.0[CoordPos::OnBoundary][CoordPos::OnBoundary] != Dimensions::Empty
+    }
+
+    /// Compares two geometry objects and returns `true` if their intersection "spatially crosses";
+    /// that is, the geometries have some, but not all interior points in common
+    ///
+    /// ```
+    /// use geo_types::{LineString, line_string, polygon};
+    /// use geo::relate::Relate;
+    ///
+    /// let line_string: LineString = line_string![(x: 85.0, y: 194.0), (x: 162.0, y: 135.0)];
+    /// let poly = polygon![
+    ///     (x: 125., y: 179.),
+    ///     (x: 110., y: 150.),
+    ///     (x: 160., y: 160.),
+    ///     (x: 125., y: 179.),
+    /// ];
+    ///
+    /// let intersection = line_string.relate(&poly);
+    /// assert_eq!(intersection.is_crosses(), true);
+    /// ```
+    /// # Notes
+    /// - If any of the following do not hold, the function will return `false`:
+    ///     - The intersection of the interiors of the geometries must be non-empty
+    ///     - The intersection must have dimension less than the maximum dimension of the two input geometries
+    ///     - The intersection of the two geometries must not equal either geometry
+    /// - This predicate is **symmetric**
+    pub fn is_crosses(&self) -> bool {
+        if self.0[CoordPos::Inside][CoordPos::Outside]
+            != self.0[CoordPos::Outside][CoordPos::Inside]
+            && self.0[CoordPos::Inside][CoordPos::Outside] != Dimensions::Empty
+            && self.0[CoordPos::Outside][CoordPos::Inside] != Dimensions::Empty
+        {
+            // input geom dimensions differ and neither are empty: proceed to calculating relationships between dimensions
+            let ii = self.0[CoordPos::Inside][CoordPos::Inside];
+            let ie = self.0[CoordPos::Inside][CoordPos::Outside];
+            let ei = self.0[CoordPos::Outside][CoordPos::Inside];
+            let a_lessthan_b = ii == ie;
+            let a_greaterthan_b = ii == ei;
+            // something has gone horribly wrong
+            assert!(a_lessthan_b != a_greaterthan_b);
+            if a_lessthan_b {
+                // [T*T******]
+                return self.0[CoordPos::Inside][CoordPos::Inside] != Dimensions::Empty
+                    && self.0[CoordPos::Inside][CoordPos::Outside] != Dimensions::Empty;
+            }
+            if a_greaterthan_b {
+                // [T*****T**]
+                return self.0[CoordPos::Inside][CoordPos::Inside] != Dimensions::Empty
+                    && self.0[CoordPos::Outside][CoordPos::Inside] != Dimensions::Empty;
+            }
+        } else {
+            // dimensions are equal: this is only permitted for line / line
```

Comment:
Can you clarify what you mean by `permitted` here? I understand there is a special case in the definition of `is_crosses` for line/line geometries, but do you mean that the previous `if` check precludes this block from executing on `line` /`line` geometries or that if we are in this block, anything that is not line/line will return false?

Reply:
My understanding is that the previous check is supposed to preclude this block from executing. That is, it's (obviously) possible to check this predicate on any geometry combination, but the result will be wrong (or ill-defined?) for any combination that isn't line / line as other geometries with equal dimensions cannot "cross".